trigger/src/main/scala/core/scheduler/JobScheduler.scala

```diff
@@ -0,0 +1,67 @@
+package core.scheduler
+
+import java.util.concurrent
+import java.util.concurrent.atomic.AtomicBoolean
+
+import core.persistance._
+import core.scheduler.executors.Executors
+import core.scheduler.sensors.Sensors
+import core.scheduler.utilities.{SchedulerConfig, SensorsConfig}
+
+import scala.collection.mutable
+import scala.concurrent.{ExecutionContext, ExecutionContextExecutor, Future}
+import scala.util.{Failure, Success}
+
+class JobScheduler(sensors: Sensors, executors: Executors, jobInstanceRepository: JobInstanceRepository) {
+  private val HEART_BEAT: Int = SchedulerConfig.getHeartBeat
+  val NUM_OF_PAR_TASKS: Int = SchedulerConfig.getMaxParallelJobs
+
+  private implicit val executionContext: ExecutionContextExecutor =
+    ExecutionContext.fromExecutor(concurrent.Executors.newFixedThreadPool(SensorsConfig.getThreadPoolSize))
+
+  private val isIsManagerRunning = new AtomicBoolean(false)
+  private var runningScheduler: Future[Unit] = Future.successful((): Unit)
+  private var runningSensors = Future.successful((): Unit)
+  private var runningEnqueue = Future.successful((): Unit)
+  private val runningJobs = mutable.Map.empty[Long, Future[Unit]]
+
+  def startManager(): Unit = {
+    isIsManagerRunning.set(true)
+    runningScheduler =
+      Future {
+        while(isIsManagerRunning.get()) {
+          removeFinishedJobs()
+          if(runningSensors.isCompleted)
```

Comment:
It is already nicely decomposed, but maybe follow the same pattern as `removeFinishedJobs` would improve maintainability. 

For instance, adding to new methods, `checkCompleteSensors` and `checkResourceAvailability` and then move the checks for job complete and space in the running queue to those methods. 

The final code would then be:

`while (isManagerRunning.get) {
   removeFinishedJobs()
   checkCompleteSensors()
   checkInfrastructureAvailability()
}
`

Reply:
Good idea, I have changed it.